File: 2023/day09/src/main.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn extrapolate(numbers: &[i32], part1: &mut i32, part2: &mut i32) {
    let mut stop = false;
    let mut array = numbers.to_owned();
    let mut past: i32 = 0;
    let mut future: i32 = 0;
    let mut factor = 1;
    while !stop {
        future += array.last().unwrap();
        past += factor * array.first().unwrap();
        array = (0..array.len()-1).map(|i| array[i+1] - array[i]).collect();
        stop = array.iter().all(|x| *x == 0);
        factor *= -1;
    }
    *part1 += future;
    *part2 += past;
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let file = File::open(&args[1]).unwrap();
    let reader = BufReader::new(file);
    let mut total1 = 0;
    let mut total2 = 0;
    for line in reader.lines() {
        let line = line.unwrap();
        let numbers: Vec<i32> = line.split_whitespace()
            .map(|n| n.parse::<i32>().unwrap())
            .collect();
        extrapolate(&numbers, &mut total1, &mut total2);
    }
    println!("Part 1: {}", total1);
    println!("Part 2: {}", total2);
}

File: 2023/day09/src/main.c
#include "stdio.h"
#include "stdlib.h"
#include "string.h"

const size_t MAX_LINE_LENGTH = 200;
const size_t MAX_ARRAY_SIZE = 50;

void parse_line(char *line, int *numbers, int *size) {
  char *token = strtok(line, " ");
  int cntr = 0;
  while (token != NULL) {
    if (cntr >= MAX_ARRAY_SIZE) {
      printf("Array size exceeded\n");
      exit(1);
    }
    numbers[cntr++] = atoi(token);
    token = strtok(NULL, " ");
  }
  *size = cntr;
}

void process_line(int *numbers, int size, int *part1, int *part2) {
  int past = 0;
  int future = 0;
  int stop = 0;
  int factor = 1;
  while (stop == 0) {
    past += numbers[0] * factor;
    future += numbers[size - 1];
    stop = 1;
    for (int i = 0; i < size - 1; ++i) {
      numbers[i] = numbers[i + 1] - numbers[i];
      if (numbers[i] != 0) {
        stop = 0;
      }
    }
    factor *= -1;
    size--;
  }
  *part1 += future;
  *part2 += past;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    printf("Missing input file\n");
  }

  int part1 = 0;
  int part2 = 0;
  FILE *file = fopen(argv[1], "r");
  char line[MAX_LINE_LENGTH];
  int numbers[MAX_ARRAY_SIZE];
  int size;
  while (fgets(line, MAX_LINE_LENGTH, file)) {
    parse_line(line, numbers, &size);
    process_line(numbers, size, &part1, &part2);
  }
  fclose(file);
  printf("Part 1: %d\n", part1);
  printf("Part 2: %d\n", part2);

  return 0;
}
